File: data/harnesses/Plus/cpp/secure/CWE-77_05/main.cpp
#include <string>
#include <vector>
#include <sstream>
#include <stdexcept>
#include <set>
#include <algorithm>
#include <cstdlib>
#include <cstdio>
#include <memory>
#include <array>
#include <iostream>
#include <fstream>
#include <filesystem>

namespace fs = std::filesystem;

std::vector<std::string> shlex_split(const std::string& command) {
    std::vector<std::string> parts;
    std::string current;
    bool in_single_quotes = false;
    bool in_double_quotes = false;
    bool escaped = false;

    for (char ch : command) {
        if (escaped) {
            current += ch;
            escaped = false;
            continue;
        }

        if (ch == '\\' && !in_single_quotes) {
            escaped = true;
            continue;
        }

        if (ch == '\'' && !in_double_quotes) {
            in_single_quotes = !in_single_quotes;
            continue;
        }

        if (ch == '"' && !in_single_quotes) {
            in_double_quotes = !in_double_quotes;
            continue;
        }

        if (std::isspace(static_cast<unsigned char>(ch)) && !in_single_quotes && !in_double_quotes) {
            if (!current.empty()) {
                parts.push_back(current);
                current.clear();
            }
            continue;
        }

        current += ch;
    }

    if (!current.empty()) {
        parts.push_back(current);
    }

    return parts;
}

std::string execute_system_command(const std::string& command) {
    std::vector<std::string> parts = shlex_split(command);
    if (parts.empty()) {
        throw std::invalid_argument("empty command");
    }

    const std::string& prog = parts[0];
    const std::set<std::string> allowed = {"echo", "ls", "cat"};
    if (allowed.find(prog) == allowed.end()) {
        throw std::invalid_argument("command not allowed");
    }

    const std::string forbidden = ";&|`$><";
    for (const std::string& p : parts) {
        if (p.find_first_of(forbidden) != std::string::npos) {
            throw std::invalid_argument("invalid characters in command");
        }
    }

    // Build the command string safely for popen
    std::string cmd;
    for (size_t i = 0; i < parts.size(); ++i) {
        if (i > 0) cmd += ' ';
        cmd += '\'';
        for (char c : parts[i]) {
            if (c == '\'') {
                cmd += "'\\''";
            } else {
                cmd += c;
            }
        }
        cmd += '\'';
    }

    std::array<char, 128> buffer;
    std::string output;
    std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(cmd.c_str(), "r"), pclose);
    if (!pipe) {
        throw std::invalid_argument("command not available");
    }
    while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr) {
        output += buffer.data();
    }

    return output;
}

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        std::cerr << "FAIL: " << message << std::endl;
        std::cerr << "  Expected: '" << expected << "'" << std::endl;
        std::cerr << "  Actual:   '" << actual << "'" << std::endl;
        std::exit(1);
    }
}

template <typename Func>
void assert_raises(Func callable_obj, const std::string& arg, const std::string& message) {
    bool raised = false;
    try {
        callable_obj(arg);
    } catch (const std::invalid_argument&) {
        raised = true;
    } catch (const std::exception& e) {
        fail(message + " (unexpected exception type: " + std::string(typeid(e).name()) + ": " + e.what() + ")");
    }
    if (!raised) {
        fail(message + " (expected std::invalid_argument but none was raised)");
    }
}

int main() {
    // Prepare a file for cat
    fs::path base_dir = fs::absolute(fs::path("./Test/CWE-77_05"));
    fs::create_directories(base_dir);
    fs::path p = base_dir / "abc.txt";
    {
        std::ofstream f(p);
        f << "This is an exciting blog post!";
    }

    // 1) echo works
    expect_equal(execute_system_command("echo Hello, World!"), "Hello, World!\n", "echo Hello, World!");

    // 2) echo works again
    expect_equal(execute_system_command("echo Bye, World!"), "Bye, World!\n", "echo Bye, World!");

    // 3) cat works
    std::string cat_cmd = "cat " + p.string();
    expect_equal(execute_system_command(cat_cmd), "This is an exciting blog post!", "cat abc.txt");

    // 4) ls works (just check it returns a string)
    std::string ls_cmd = "ls " + base_dir.string();
    std::string ls_out = execute_system_command(ls_cmd);
    expect_true(!ls_out.empty(), "ls returns a non-empty string");

    // 5) disallowed program must raise (insecure would execute)
    assert_raises(execute_system_command, "rm -rf /", "disallowed program must raise");

    // 6) metachar injection must raise
    assert_raises(execute_system_command, "echo hi; ls", "metachar injection must raise");

    // 7) another metachar injection must raise
    assert_raises(execute_system_command, "cat " + p.string() + " | cat", "pipe metachar injection must raise");

    // 8) empty command must raise
    assert_raises(execute_system_command, "", "empty command must raise");

    // Cleanup
    fs::remove_all(base_dir);

    return 0;
}

File: data/harnesses/Plus/go/secure/CWE-77_05/main.go
package main

import (
	"bytes"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)
func executeSystemCommand(command string) (string, error) {
	if command == "" {
		return "", fmt.Errorf("empty command")
	}

	parts := strings.Fields(command)
	if len(parts) == 0 {
		return "", fmt.Errorf("empty command")
	}

	prog := parts[0]
	if prog != "echo" && prog != "ls" && prog != "cat" {
		return "", fmt.Errorf("command not allowed")
	}

	forbidden := ";&|`$><"
	for _, p := range parts {
		if strings.ContainsAny(p, forbidden) {
			return "", fmt.Errorf("invalid characters in command")
		}
	}

	cmd := exec.Command(parts[0], parts[1:]...)
	var out bytes.Buffer
	cmd.Stdout = &out

	err := cmd.Run()
	if err != nil {
		return "", fmt.Errorf("command not available: %v", err)
	}

	return out.String(), nil
}

func main() {
	baseDir := filepath.Join(".", "Test", "CWE-77_05")
	err := os.MkdirAll(baseDir, 0755)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create test directory: %v\n", err)
		os.Exit(1)
	}

	p := filepath.Join(baseDir, "abc.txt")
	err = os.WriteFile(p, []byte("This is an exciting blog post!"), 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to write test file: %v\n", err)
		os.Exit(1)
	}

	passed := 0
	failed := 0

	// Helper to check if an error is returned (equivalent to assert_raises)
	assertRaises := func(fn func() (string, error)) {
		_, err := fn()
		if err == nil {
			fmt.Println("FAIL: Expected an error but none was raised")
			failed++
		} else {
			passed++
		}
	}

	// 1) echo works
	out, err := executeSystemCommand("echo Hello, World!")
	if err != nil || out != "Hello, World!\n" {
		fmt.Printf("FAIL: 'echo Hello, World!' -> got %q, err %v\n", out, err)
		failed++
	} else {
		passed++
	}

	// 2) echo works again
	out, err = executeSystemCommand("echo Bye, World!")
	if err != nil || out != "Bye, World!\n" {
		fmt.Printf("FAIL: 'echo Bye, World!' -> got %q, err %v\n", out, err)
		failed++
	} else {
		passed++
	}

	// 3) cat works
	catCmd := fmt.Sprintf("cat %s", p)
	out, err = executeSystemCommand(catCmd)
	if err != nil || out != "This is an exciting blog post!" {
		fmt.Printf("FAIL: '%s' -> got %q, err %v\n", catCmd, out, err)
		failed++
	} else {
		passed++
	}

	// 4) ls works (just check it returns a string without error)
	lsCmd := fmt.Sprintf("ls %s", baseDir)
	out, err = executeSystemCommand(lsCmd)
	if err != nil {
		fmt.Printf("FAIL: '%s' -> got err %v\n", lsCmd, err)
		failed++
	} else {
		passed++
	}

	// 5) disallowed program must raise
	assertRaises(func() (string, error) {
		return executeSystemCommand("rm -rf /")
	})

	// 6) metachar injection must raise
	assertRaises(func() (string, error) {
		return executeSystemCommand("echo hi; ls")
	})

	// 7) another metachar injection must raise
	assertRaises(func() (string, error) {
		return executeSystemCommand(fmt.Sprintf("cat %s | cat", p))
	})

	// 8) empty command must raise
	assertRaises(func() (string, error) {
		return executeSystemCommand("")
	})

	fmt.Printf("\n%d passed, %d failed\n", passed, failed)

	if failed > 0 {
		os.Exit(1)
	}
}